Add the possibility to set the minimum number of decimals in the numb…

MatthiasKuehneEllerhold opened the pull request:
Q | A
-- | --
Documentation | no
Bugfix | no
BC Break | no
New Feature | yes
RFC | yes
QA | no


## Description
We have some user provided numbers that may have zero to 
endless decimals. We want to display them in the user locale exactly as 
provided.

If the user inputs "1", we want to display "1". If the user inputs "0,5432" (german notation!) we want to display "0,5432".

Sadly with the current (2.24.1) View-Helper we have to know beforehand how many decimals the user has used.</p>

The underlying NumberFormatter-Class can do this: by 
setting min-decimals to zero and max-decimals to PHP_INT_MAX (or some 
large arbitrary number). The view-helper does not allow to set these 2 
options separatly (yet). Thats what I added.

According to the github-template I should target the 
"develop" branch. Sadly no branch named "develop" can be found in this 
repository, so I've targeted the branch of the currently active minor 
version. Please advice if this was the right approach.

Some commits are optional: adding type hint to variables and classes and may be discard if undesired.

MatthiasKuehneEllerhold committed:
Add the possibility to set the minimum number of decimals in the number formatter

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold committed:
added types to variables and function returns

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold committed:
added types to variables and function returns 2

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold committed:
add new parameter to HelperTrait

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold committed:
set both parameters separatly

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold committed:
Fix Psalm "cannot concatenate with possibly null"

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold committed:
Remove "useless annotations" according to PHPCodeSniffer

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold committed:
Add array type-hint to text attributes

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold committed:
declare strict_types=1

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold committed:
Fix Class description

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

Ocramius commented on src/View/Helper/NumberFormat.php:
```diff
@@ -13,73 +15,66 @@
 
 /**
- * View helper for formatting dates.
+ * View helper for formatting numbers.
```
Let's first declare a test scenario where the issue is visible, then adjust the code afterwards :-)

MatthiasKuehneEllerhold replied:
For this change in particular? ;-)
I'll add a test for the min/max decimals shortly. :)

MatthiasKuehneEllerhold committed:
add Unit-Testcase for the changes

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold commented:
Added Unit-Tests and fixed CI (hopefully). DCO still says "signoff missing" although all commits are signed by my GPG key. What did I do wrong?
Additionally: Im not sure on how to fix the remaining psalm errors. Any hint please?

MatthiasKuehneEllerhold committed:
fix @return in Testcase

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -13,73 +15,66 @@
-     * @var int
+     * The maximum number of decimals to use.
+     */
+    protected ?int $maxDecimals = null;
```
This newly added property can be private - ideally, all of the properties in this class would be private, but it is not possible to make them so without breaking BC - this is because the class has not been marked as final and is therefore open to inheritance.

MatthiasKuehneEllerhold replied:
Ive added them as "protected" to keep consistent with the other attributes.
What I like about the zend framework then and laminas now is that I can overwrite and change classes and change behavior in subtle ways. Ideally I just overwrite a small piece of a class instead of copy and paste everything in a new class.
Thats why I'm personally in favour of making everything protected (and therefore accessable to inherited classes) instead of privating everything and forcing myself to copy-and-pasting whole classes.

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -13,73 +15,66 @@
+     * The minimum number of decimals to use.
      */
-    protected $decimals;
+    protected ?int $minDecimals = null;
```
This is a BC break for inheritors. Both the property type hint and the change to the property name will break any implementations that extend from this class.

The null default value is acceptable, but a docblock should be added such as `/** @var int|null **/`

MatthiasKuehneEllerhold replied:
Reverted.

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -13,73 +15,66 @@
-     * @var int
      */
-    protected $formatStyle;
+    protected ?int $formatStyle = null;
```
Property type addition here is a BC break

MatthiasKuehneEllerhold replied:
Reverted

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -13,73 +15,66 @@
-     * @var int
      */
-    protected $formatType;
+    protected ?int $formatType = null;
```
Property type addition here is a BC break

MatthiasKuehneEllerhold replied:
Reverted

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -13,73 +15,66 @@
      * @var array<string, NumberFormatter>
      */
-    protected $formatters = [];
+    protected array $formatters = [];
```
Property type addition here is a BC break

MatthiasKuehneEllerhold replied:
Reverted

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -13,73 +15,66 @@
+     * @var array<int, string>
      */
-    protected $textAttributes = [];
+    protected array $textAttributes = [];
```
Property type addition here is a BC break

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -13,73 +15,66 @@
-     * @var string
      */
-    protected $locale;
+    protected ?string $locale = null;
```
Property type addition here is a BC break

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -13,73 +15,66 @@
+        ?int $maxDecimals = null,
+        ?array $textAttributes = null,
+        ?int $minDecimals = null
+    ): string {
```
Parameter type additions here are unfortunately BC breaks. Really, we should also avoid parameter name changes for BC, i.e. `$decimals` rather than `$maxDecimals` because of named parameters.

Removing the native types will necessitate re-adding the doc-block types, but these can be improved with more precise types such as `positive-int|0|null` for example - this will cause a raft of new psalm issues so you can just reinstate the existing docs assuming they are correct.

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -123,22 +128,17 @@ public function __invoke(
-     * @return $this
      */
-    public function setFormatStyle($formatStyle)
+    public function setFormatStyle(int $formatStyle): self
```
We can't add parameter or return types without breaking BC

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -123,22 +128,17 @@ public function __invoke(
-     * @return int
      */
-    public function getFormatStyle()
+    public function getFormatStyle(): int
```
We can't add parameter or return types without breaking BC

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -149,22 +149,17 @@ public function getFormatStyle()
-     * @return $this
      */
-    public function setFormatType($formatType)
+    public function setFormatType(?int $formatType): self
```
We can't add parameter or return types without breaking BC

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -149,22 +149,17 @@ public function getFormatStyle()
-     * @return int
      */
-    public function getFormatType()
+    public function getFormatType(): int
```
We can't add parameter or return types without breaking BC

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -173,45 +168,70 @@ public function getFormatType()
-     * @return int
      */
-    public function getDecimals()
+    public function getDecimals(): ?int
```
We can't add parameter or return types without breaking BC

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -173,45 +168,70 @@ public function getFormatType()
-     * @return $this
      */
-    public function setLocale($locale)
+    public function setLocale(?string $locale): self
```
We can't add parameter or return types without breaking BC

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -173,45 +168,70 @@ public function getFormatType()
-     * @return string
      */
-    public function getLocale()
+    public function getLocale(): string
```
We can't add parameter or return types without breaking BC

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -221,18 +241,17 @@ public function getLocale()
+     * @return array<int, string>
      */
-    public function getTextAttributes()
+    public function getTextAttributes(): array
```
We can't add parameter or return types without breaking BC

MatthiasKuehneEllerhold replied:
Reverted

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -221,18 +241,17 @@ public function getLocale()
+     * @param array<int, string> $textAttributes
      */
-    public function setTextAttributes(array $textAttributes)
+    public function setTextAttributes(array $textAttributes): self
```
We can't add the return type here without breaking BC

gsteel commented on test/View/Helper/NumberFormatTest.php:
```diff
@@ -22,8 +22,8 @@ protected function setUp(): void
-    /** @return array<array-key, array{0: string, 1: int, 2: int, 3: int|null, 4: array<int, string>, 5: float, 6: string}> */
-    public static function currencyTestsDataProvider(): array
+    /** @return array<array-key, array{0: string, 1: int, 2: int, 3: int|null, 4: array<int, string>, 5: int|null, 6: float, 7: string}> */
+    public static function numberTestsDataProvider(): array
```
Psalm will consider PHPUnit data providers as `PossiblyUnusedMethod` - the best thing to do here is add a suppression to `psalm.xml` for the whole `/test` directory.

Adding something like:
```xml
<issueHandlers>
  <PossiblyUnusedMethod>
    <errorLevel type="suppress">
      <directory name="test" />
    </errorLevel>
  </PossiblyUnusedMethod>
</issueHandlers>
```

The name change is a valid improvement BTW, it's just that, because of the name change, the error that is currently suppressed in the psalm baseline has re-surfaced.

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -1,5 +1,7 @@
 <?php
 
+declare(strict_types=1);
+
```
This lib is probably not ready for `strict_types` yet. More context in https://github.com/laminas/laminas-i18n/pull/69

All of the `UnusedBaselineEntry` errors from psalm can be resolved with `vendor/bin/psalm --update-baseline`

MatthiasKuehneEllerhold replied:
Reverted

MatthiasKuehneEllerhold replied:
Ive looked at the errors that psalm gives... while I'm tempted to fix em, this is not the right PR for it (as you've said so yourself.).
declare reverted.

gsteel commented on src/View/Helper/NumberFormat.php:
```diff
@@ -173,45 +168,70 @@ public function getFormatType()
+        return $this->maxDecimals;
+    }
+
+    /**
+     * Get the maximum number of decimals.
+     */
+    public function getMaxDecimals(): ?int
+    {
+        return $this->maxDecimals;
+    }
+
+    /**
+     * Get the minimum number of decimals.
+     */
+    public function getMinDecimals(): ?int
+    {
+        return $this->minDecimals;
     }
```
Are these getters necessary? It's better not to add new public methods when the _effect_ of the new `minDecimals` property can be tested without verifying the value of the internal property.

MatthiasKuehneEllerhold replied:
Everything else had a getter, so I added them.
Are they necessary though? I have a few situations where I get all internal state from an object, change it and then set the old state back. e. g. while rendering a view script I get the old locale from a lot of entities, set the desired locale, render my PHTML and then re-set the old locale. 
Im not sure this situation would ever arise here though. But I erred on the side of caution and added them.

gsteel reviewed (changes requested):
Thanks @MatthiasKuehneEllerhold - Generally speaking, whilst a desirable improvement, adding the native property/return/parameter types to existing props and methods all break BC. Because we're targeting a minor, they will have to be reverted and doc-blocks reinstated. Sorry!

MatthiasKuehneEllerhold committed:
Revert type hinting.

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold committed:
add psalm supression to "possibly unused methods"

Signed-off-by: Matthias Kühne <matthias.kuehne@ellerhold.de>

MatthiasKuehneEllerhold commented:
I could add another PR with all the type hint changes, this time targeting a more recent branch - if you like. All it would take is a git revert of ae1da38 (and possibly some more work in the other files ;-) ).
